recourse_methods/test_face_method.py

```diff
@@ -0,0 +1,264 @@
+        # weight.
+        # Index 1 is chosen second because it is a candidate and its index is
+        # lower than the alternative it ties with.
+        expected_candidate_indices = np.array([3, 1])
+
+        candidate_indices = face_method.FACE._get_k_best_candidate_indices(
+            distances, candidate_indices, k_candidates
+        )
+
+        np.testing.assert_equal(candidate_indices, expected_candidate_indices)
+
+    def test_get_k_best_candidate_indices_failure(self):
+        # There are four points, but only the last three are candidates.
+        distances = np.array([1, np.inf, np.inf, 1])
+        candidate_indices = np.array([1, 2, 3])
+        k_candidates = 2
+
+        # Index 3 is chosen first because it is a candidate and has minimal
+        # weight.
+        # No other indices can be chosen -- the other candidates have infinite
+        # weight.
+        expected_candidate_indices = np.array([3])
+
+        candidate_indices = face_method.FACE._get_k_best_candidate_indices(
+            distances, candidate_indices, k_candidates
+        )
+
+        np.testing.assert_equal(candidate_indices, expected_candidate_indices)
+
+    @mock.patch(
+        "recourse_methods.face_method.recourse_adapter.RecourseAdapter",
+        autospec=True,
+    )
+    def test_get_paths_from_indices(self, mock_adapter):
+        # Test reconstructing the path POI -> 1 -> 2
+        dataset = pd.DataFrame({"a": [1, 2, 3], "label": [0, 1, 1]})
+        poi = pd.Series([0], index=["a"])
+        target_indices = [2]
+        predecessors = [
+            -9999,  # there is no path to point 0
+            -1,  # point 1 is connected to the POI
+            1,  # the path to point 2 runs through point 1
+        ]
+
+        # The adapter is the identity function
+        mock_adapter.label_column = "label"
+        mock_adapter.transform.side_effect = lambda x: x
+
+        mock_self = mock.Mock(spec=["adapter", "dataset"])
+        mock_self.adapter = mock_adapter
+        mock_self.dataset = dataset
+
+        # The expected path goes through the POI, the point at index 1, and the
+        # point at index 2.
+        expected_paths = [pd.DataFrame({"a": [0, 2, 3]})]
+
+        paths = face_method.FACE._get_paths_from_indices(
+            mock_self, target_indices, poi, predecessors
+        )
+
+        self.assertEqual(len(paths), len(expected_paths))
+
+        pd.testing.assert_frame_equal(paths[0], expected_paths[0])
+
+    @mock.patch(
+        "recourse_methods.face_method.recourse_adapter.RecourseAdapter",
+        autospec=True,
+    )
+    @mock.patch("recourse_methods.face_method._get_edge_weight")
+    def test_append_new_point(self, mock_get_edge_weight, mock_adapter):
+        distance_threshold = 1
+        dataset = pd.DataFrame({"a": [0, 1, 2], "label": [0, 1, 1]})
+
+        # The graph to append a new point to
+        graph = sparse.csr_array(
```

Comment:
Can you please simplify this test example? The initial graph and clarify the input dataset "a" - is that a dataset with a single 3-dimensional point, or three 1-dimensional point? It's hard to discern from the setup.

Reply:
From offline discussion with @jakeval , this test function currently
* calculates the distance from the new point to every other point
* calculates the weight from the distance
* replaces over-threshold distances with 0-weight
* adds the distance vector to the original graph as a column and a row

Would you please split these out into individual test functions? I think that'll simplify this and make this easier to maintain. Thank you!

Reply:
The function is now called _add_point_to_graph() and is no longer directly tested. Instead, it uses the helper functions _calculate_weight_vector() and _add_edges_to_graph() which are unit tested.